add support for `#[defmt(transparent)]`

vic1707 opened the pull request:
Fixes: #934, inspired by https://defmt.ferrous-systems.com/format#newtypes

Here's a proposition implementing a new attribute `#[defmt(transparent)]` for newtypes and enums (where each variant has only one field).

I tested using
```rs
#[derive(defmt::Format)]
#[defmt(transparent)]
struct Unnamed(u8);
#[derive(defmt::Format)]
#[defmt(transparent)]
struct Named {
    foo: u8,
}
#[derive(defmt::Format)]
#[defmt(transparent)]
enum Mixed {
    Unnamed(u8),
    Named { foo: u8 },
    StructUnnamed(Unnamed),
    StructNamed { named: Named },
}
```

<details>
<summary> generated code </summary>

```rs
#[defmt(transparent)]
struct Unnamed(u8);
impl defmt::Format for Unnamed {
    fn format(&self, f: defmt::Formatter) {
        self.0.format(f)
    }
}
#[defmt(transparent)]
struct Named {
    foo: u8,
}
impl defmt::Format for Named {
    fn format(&self, f: defmt::Formatter) {
        self.foo.format(f)
    }
}
#[defmt(transparent)]
enum Mixed {
    Unnamed(u8),
    Named { foo: u8 },
    StructUnnamed(Unnamed),
    StructNamed { named: Named },
}
impl defmt::Format for Mixed {
    fn format(&self, f: defmt::Formatter) {
        match &self {
            Self::Unnamed { 0: inner } => inner.format(f),
            Self::Named { foo: inner } => inner.format(f),
            Self::StructUnnamed { 0: inner } => inner.format(f),
            Self::StructNamed { named: inner } => inner.format(f),
        }
    }
}
```

</details>

Waiting for suggestion's approbation before finishing the PR (docs, changelog, generic super trait requirements etc).

vic1707 commented:
Just added the support for generic in the derive

---
I also took the liberty to add #[derive(defmt::FormatTransparent)] because I intend to use that feature with [auto_enums](https://github.com/taiki-e/auto_enums) which doesn't appear to support added attributes.
This is completely selfish and I will understand if you ask me to remove it, I'll just go to auto_enums and try to add support for attributes there.

jonathanpallant commented:
I think it makes sense for new-types (`struct Thing(u32)`) but I don't know that it makes sense for enums, because you're throwing away the discriminant.

vic1707 commented:
I added it for 2 reasons,
- it was quick & easy, users could want it 
- I want to use it in conjunction with `auto_enums` so I don't care at all about the enum itself 🙃 

But the final decision is made by you 👍

jonathanpallant commented:
ping @Veykril on this one

vic1707 commented on macros/src/lib.rs:
```diff
@@ -40,6 +40,13 @@ pub fn format(input: TokenStream) -> TokenStream {
     derives::format::expand(input)
 }
 
+#[proc_macro_derive(FormatTransparent)]
+#[proc_macro_error]
+pub fn format_transparent(input: TokenStream) -> TokenStream {
+    let input = syn::parse_macro_input!(input as syn::DeriveInput);
+    derives::format::expand_transparent(input)
+}
```
from https://github.com/knurling-rs/defmt/pull/937#issuecomment-2660479379
> I also took the liberty to add #[derive(defmt::FormatTransparent)] because I intend to use that feature with [auto_enums](https://github.com/taiki-e/auto_enums) which doesn't appear to support added attributes.
This is completely selfish and I will understand if you ask me to remove it, I'll just go to `auto_enums` and try to add support for attributes there.

After thinking about it a bit more I think the support should be on `auto_enums` side rather than exposing a second `defmt` derive. I'll let you be the judge 👍

vic1707 commented:
tests and support for `crate` attribute are being worked on, I'll wrap it up asap, thanks for the review

vic1707 commented:
Updated branch from main
rebased on https://github.com/knurling-rs/defmt/pull/957
added tests and remove the `FormatTransparent` (I can reintroduce it quickly, waiting for your decision)
Everything should be good to go for a new review 🙃
